File: scala-practice/app/src/main/scala/io/wcygan/algorithms/search/BinarySearch.scala
package io.wcygan.algorithms.search

import scala.annotation.tailrec

object BinarySearch {
  @tailrec
  def search[T: Ordering](value: T, arr: IndexedSeq[T]): Option[T] = {
    if (arr.isEmpty) return None
    if (arr.length == 1 && !arr(0).equals(value)) return None
    val mid = arr(arr.length / 2)
    if (mid.equals(value)) Some(mid)
    else {
      val (lo, hi) = if (Ordering[T].gt(mid, value)) {
        (0, arr.length / 2)
      } else {
        (arr.length / 2, arr.length)
      }
      search(value, arr.slice(lo, hi))
    }
  }
}

File: java-practice/src/main/java/io/wcygan/algorithms/miscellaneous/BinarySearch.java
package io.wcygan.algorithms.miscellaneous;

import java.util.List;

public class BinarySearch {
    public static <T extends Comparable<T>> int search(List<T> list, T toFind) {
        int lo = 0;
        int hi = list.size() - 1;

        while (lo <= hi) {
            int mid = (lo + hi) / 2;

            if (list.get(mid).compareTo(toFind) == 0) {
                return mid;
            } else if (list.get(mid).compareTo(toFind) < 0) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        return -1;
    }
}
